java-compiler-testing/src/main/java/io/github/ascopes/jct/workspaces/Workspace.java

```diff
@@ -0,0 +1,108 @@
+package io.github.ascopes.jct.workspaces;
+
+import static java.util.Objects.requireNonNull;
+
+import io.github.ascopes.jct.filemanagers.ModuleLocation;
+import io.github.ascopes.jct.workspaces.impl.BasicPathWrapperImpl;
+import io.github.ascopes.jct.workspaces.impl.RamDirectory;
+import io.github.ascopes.jct.workspaces.impl.TempDirectory;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import javax.tools.JavaFileManager.Location;
+import org.apiguardian.api.API;
+import org.apiguardian.api.API.Status;
+
+// TODO(ascopes): interface time.
+public class Workspace implements AutoCloseable {
```

Comment:
- [ ] Do I want to use an interface for this? The impl can be exposed via a static method on the interface.